[quant][graph] Add _choose_qparams function for graph mode

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35366 [quant][graph] Add pass to insert quant dequant for dynamic quantization
* #35325 [quant][graph] Update dynamic quant tests to use new qconfig
* #35265 [quant][graph] Add a new observer type for dynamic quantization
* **#35235 [quant][graph] Add _choose_qparams function for graph mode**

Summary: For dynamic quantization in graph mode, we need an operator that returns the qparams of the tensor
similar to the linear_dynamic quantized op

Test Plan:
python test/test_quantized_tensor.py TestQuantizedTensor.test_choose_qparams

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20608793](https://our.internmc.facebook.com/intern/diff/D20608793)

supriyar committed:
[quant][graph] Add _choose_qparams function for graph mode

Summary: For dynamic quantization in graph mode, we need an operator that returns the qparams of the tensor
similar to the linear_dynamic quantized op

Test Plan:
python test/test_quantized_tensor.py TestQuantizedTensor.test_choose_qparams

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/utils/wrap_outputs.h:
```diff
@@ -185,4 +185,13 @@ inline PyObject* wrap(at::IntArrayRef list) {
 }
+
+inline PyObject* wrap(std::tuple<float, int64_t> tensors) {
+  auto r = THPObjectPtr{PyTuple_New(5)};
```
Shouldn't this be `2`?

supriyar replied:
Yeah, typo. Fixed it.

supriyar committed:
Update on "[quant][graph] Add _choose_qparams function for graph mode"


Summary: For dynamic quantization in graph mode, we need an operator that returns the qparams of the tensor
similar to the linear_dynamic quantized op

Test Plan:
python test/test_quantized_tensor.py TestQuantizedTensor.test_choose_qparams

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -218,5 +219,22 @@ bool quantized_equal(const Tensor& self, const Tensor& other) {
 }
 
+/* Calculate the quantization params for the activation tensor */
+std::tuple<double, int64_t> _choose_qparams(const Tensor& self, bool reduce_range) {
```
we'll need to have another function for per channel quantization as well until we expose Quantizer to python, maybe rename this to `_choose_qparams_per_tensor` now?

supriyar replied:
Do we need per-channel quantization for activations though?

jerryzh168 replied:
probably not, but this is still a `_choose_qparams` only for per tensor quantization, I feel it's better to be more explicit.

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -218,5 +219,22 @@ bool quantized_equal(const Tensor& self, const Tensor& other) {
+        /*max=*/x_max,
+        /*qmin=*/0,
+        /*qmax=*/255,
+        reduce_range=reduce_range);
```
really? can you pass keyword arguments in C++?

supriyar replied:
Ah no. It was supposed to be a comment. I'll update the PR

supriyar committed:
Update on "[quant][graph] Add _choose_qparams function for graph mode"


Summary: For dynamic quantization in graph mode, we need an operator that returns the qparams of the tensor
similar to the linear_dynamic quantized op

Test Plan:
python test/test_quantized_tensor.py TestQuantizedTensor.test_choose_qparams

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Looks good

supriyar committed:
Update on "[quant][graph] Add _choose_qparams function for graph mode"


Summary: For dynamic quantization in graph mode, we need an operator that returns the qparams of the tensor
similar to the linear_dynamic quantized op

Test Plan:
python test/test_quantized_tensor.py TestQuantizedTensor.test_choose_qparams

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20608793](https://our.internmc.facebook.com/intern/diff/D20608793)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add _choose_qparams function for graph mode"


Summary: For dynamic quantization in graph mode, we need an operator that returns the qparams of the tensor
similar to the linear_dynamic quantized op

Test Plan:
python test/test_quantized_tensor.py TestQuantizedTensor.test_choose_qparams

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20608793](https://our.internmc.facebook.com/intern/diff/D20608793)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add _choose_qparams function for graph mode"


Summary: For dynamic quantization in graph mode, we need an operator that returns the qparams of the tensor
similar to the linear_dynamic quantized op

Test Plan:
python test/test_quantized_tensor.py TestQuantizedTensor.test_choose_qparams

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20608793](https://our.internmc.facebook.com/intern/diff/D20608793)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add _choose_qparams function for graph mode"


Summary: For dynamic quantization in graph mode, we need an operator that returns the qparams of the tensor
similar to the linear_dynamic quantized op

Test Plan:
python test/test_quantized_tensor.py TestQuantizedTensor.test_choose_qparams

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20608793](https://our.internmc.facebook.com/intern/diff/D20608793)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add _choose_qparams function for graph mode"


Summary: For dynamic quantization in graph mode, we need an operator that returns the qparams of the tensor
similar to the linear_dynamic quantized op

Test Plan:
python test/test_quantized_tensor.py TestQuantizedTensor.test_choose_qparams

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20608793](https://our.internmc.facebook.com/intern/diff/D20608793)

[ghstack-poisoned]
